///////////////////////////////////////////////////////////////////////////
// Copyright (C) Flowbox, Inc / All Rights Reserved
// Unauthorized copying of this file, via any medium is strictly prohibited
// Proprietary and confidential
// Flowbox Team <contact@flowbox.io>, 2014
///////////////////////////////////////////////////////////////////////////

package generated.proto.interpreter;

import "crumb.proto";



message CallPoint {
    optional int32 libraryID = 1;
    optional int32 nodeID    = 2;
}

message CallPointPath {
    repeated CallPoint calls = 1;
}

message DefPoint {
    optional int32 libraryID = 1;
    optional generated.proto.crumb.Breadcrumbs breadcrumbs = 2;
}

message Interpreter {
    message WatchPoint {
        message Add {
            message Request {
                required CallPointPath watchPoint = 1;
            }
            message Update {
                required Request request = 1;
            }
        }
        message Remove {
            message Request {
                required CallPointPath watchPoint = 1;
            }
            message Update {
                required Request request = 1;
            }
        }
        message List {
            message Request { }
            message Status {
                required Request request = 1;
                repeated CallPointPath watchPoints = 2;
            }
        }
    }
    message SetProjectID {
        message Request {
            required int32 projectID = 1;
        }
        message Update {
            required Request request = 1;
        }
    }
    message SetMainPtr {
        message Request {
            required DefPoint mainPtr = 1;
        }
        message Update {
            required Request request = 1;
        }
    }
    message Run {
        message Request { }
        message Update  { 
            required Request request = 1;
        }
    }
    message Value {
        message Request {
            required CallPointPath callPointPath = 1;
        }
        message Update {
            required CallPointPath callPointPath = 1;
            required bytes         value         = 2;
        }
    }
    message Ping {
        message Request { }
        message Status  { 
            required Request request = 1;
        }
    }
}